catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/util/impl/QueryResultPaginatorTest.java

```diff
@@ -0,0 +1,395 @@
+
+    private QueryImpl queryMock;
+
+    private QueryResultPaginator paginator;
+
+    @Before
+    public void setUp() throws Exception {
+
+        catalogFrameworkMock = mock(CatalogFramework.class);
+        QueryRequestImpl queryRequestMock = mock(QueryRequestImpl.class);
+        queryResponseMock = mock(QueryResponseImpl.class);
+        queryMock = mock(QueryImpl.class);
+        when(catalogFrameworkMock.query(any(QueryRequest.class))).thenReturn(queryResponseMock);
+        when(queryRequestMock.getQuery()).thenReturn(queryMock);
+    }
+
+    @Test
+    public void testhasNext() throws Exception {
+        setStartIndexAndPageSize(1, 1);
+        setResponses(1, 0);
+        assertThat(paginator.hasNext(), equalTo(true));
+        assertThat(paginator.hasNext(), equalTo(true));
+        assertThat(paginator.hasNext(), equalTo(true));
+        paginator.next();
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.getBufferSize(), equalTo(0));
+
+    }
+
+    @Test
+    public void testSimpleNextPage() {
+        setStartIndexAndPageSize(1, 1);
+        setResponses(2, 0);
+        assertThat(paginator.next()
+                .size(), equalTo(1));
+        assertThat(paginator.next()
+                .size(), equalTo(1));
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.getBufferSize(), equalTo(0));
+
+    }
+
+    @Test(expected = NoSuchElementException.class)
+    public void testNoSuchElementException() {
+        setStartIndexAndPageSize(1, 1);
+        setResponses(0);
+        paginator.next();
+    }
+
+    @Test
+    public void testDifferentSizePages() {
+        int pageSize = 3;
+        setStartIndexAndPageSize(1, pageSize);
+        setResponses(2, 2, 0);
+        assertThat(paginator.hasNext(), equalTo(true));
+        assertThat(paginator.next()
+                .size(), equalTo(pageSize));
+        assertThat(paginator.next()
+                .size(), equalTo(1));
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.getBufferSize(), equalTo(0));
+    }
+
+    @Test
+    public void testIndexingAcrossPages() {
+        int pageSize = 5;
+        setStartIndexAndPageSize(10, pageSize);
+        setResponses(5, 5, 0);
+        List<Result> pageOne = paginator.next();
+        assertThat(pageOne, hasSize(5));
+        assertThat(paginator.getIndexOffset(), equalTo(15));
+        List<Result> pageTwo = paginator.next();
+        assertThat(pageTwo, hasSize(5));
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.getBufferSize(), equalTo(0));
+        assertThat(paginator.getIndexOffset(), equalTo(20));
+
+        try {
+            paginator.next();
+        } catch (NoSuchElementException e) {
+            //do nothing
```

Comment:
If we don't expect this exception to actually happen, then we should have the test method throw it. If we are expecting it, we should make sure that we're catching it and fail if it isn't thrown